### DISCUSSION TITLE
docker下 指定端口不生效。

### DISCUSSION THREAD

**Discussion by jianjam** [1715173329/UnblockNeteaseMusic#306] [Q&A] [answered]
小白新手，环境：主路由 192.168.2.1
docker跑在 arm的N1：192.168.2.254，使用macvlan 网络跑UNM 
N1主机的80端口运行了其他东西，所以用macvlan的方式

使用这条命令创建的macvlan
> docker network create -d macvlan --subnet=192.168.2.0/24 --gateway=192.168.2.1 -o parent=eth0 -o macvlan_mode=bridge macnet

docker运行指令

> docker run -d \
--name ubmusic \
--restart=always \
--hostname ubmusic \
--net=macnet --ip=192.168.2.222 --dns=223.5.5.5 --mac-address C2:F2:9C:C5:B2:91 \
--privileged=true \
-e TZ=Asia/Shanghai \
-e ENABLE_FLAC=true \
-e ENABLE_LOCAL_VIP=true \
pan93412/unblock-netease-music-enhanced -p 80  -f 59.111.160.195

然后主路由在dnsmasq 修改hosts 把 interface.music.163.com和music.163.com指向192.168.2.222
然后运行docker之后看日志

> INFO: (app) HTTP Server running @ http://0.0.0.0:8080
INFO: (app) HTTPS Server running @ http://0.0.0.0:8081

还是监听的8080端口， pc 客户端手动设置自定义代理 192.168.2.222 端口80显然不行， 8080 端口可以使用。但是docker地址里有很多

> ERROR: (hook) A error occurred in hook.request.after when hooking http://music.163.com/eapi/batch.,
    TypeError: Cannot set properties of undefined (setting 'expireTime'),
        at /app/app.js:14291:59,
        at processTicksAndRejections (node:internal/process/task_queues:96:5),

的错误。暂时好像不影响使用。。

N1里查看运行命令是 node app.js -o bilibili kugou kuwo 。指定的-p -f 参数貌似没齐作用。 
因该是我使用方式不对吧？不知道指定-p -f -s -e 这样的参数在哪里填写
我想让其运行在80端口，这样客户端内不用设置自定义端口了。部署docker时 -p 80:8080 映射方式测试了不能用。

谢谢解答疑惑。

**Comment by 1715173329** [comment]
如果你 n1 80 端口被其他程序占用，则无法使 unblockneteasemusic 也使用同一个端口。

**Reply by jianjam** [reply]
我把主路由上80端口的程序停止之后，发现运行UNM还是监听的8080端口。
PS ，我用nondanee / UnblockNeteaseMusic ，他的docker镜像，使用相同的命令就可以部署到80端口。

**Comment by pan93412** [comment]
<img width="842" alt="截圖 2021-11-07 下午3 57 24" src="https://user-images.githubusercontent.com/28441561/140637167-e6f81137-8cf7-48e8-aaab-8975be30d268.png">

`-p` 要寫在 image 名稱之前。我剛測試了一下，應該就沒問題了。

**Reply by jianjam** [reply]
如果使用host网络，这种方式把8080映射到80端口是可以用的，但是我需要使用和主路由不同的IP 。所以使用macvlan的方式。主路由的80端口使用了其他的服务。

**Comment by pan93412** [comment, marked as answer]
我找出並修正問題了 (#307)。等等請執行

```bash
docker pull pan93412/unblock-netease-music-enhanced
```

並重新執行 `docker run` 即可使用。

**Reply by pan93412** [reply]
@jianjam 解決問題也請幫我點一下 `Mark as answer`，謝謝！

**Comment by jianjam** [comment]
这个是nondanee

> docker run -d \\
--name ubmusic \\
--restart=always \\
--hostname ubmusic \\
--net=macnet --ip=192.168.2.222 --dns=223.5.5.5 --mac-address C2:F2:9C:C5:B2:91 \\
--privileged=true \\
-e TZ=Asia/Shanghai \\
nondanee/unblockneteasemusic \\
-p 80 -e http://music.163.com -f 59.111.160.195

docker里查看监听端口是80，运行指令
![I1wkss.md.png](https://z3.ax1x.com/2021/11/07/I1wkss.md.png)

pan93412
![I1dIIK.md.png](https://z3.ax1x.com/2021/11/07/I1dIIK.md.png)
监听端口是8080.

两个node的执行指令有点区别，这个是不是影响参数传入的原因？我对node不太懂。希望能给予一点参考
